Cisco FWSM can add protocol to message 106023.

nhuff opened the pull request:
In the case of a numeric protocol cisco will log 'protocl number'
instead of a protocol string
ie.
Deny protocol 103 src outside:xxx.xxx.xxx.xxx dst inside:xxx.xxx.xxx.xxx by access-group "S_OUTSIDE_TO_INSIDE" [0x54c7fa80, 0x0]
vs
Deny tcp src outside:xxx.xxx.xxx.xxx/48821 dst inside:xxx.xxx.xxx.xxx/25 by access-group "S_OUTSIDE_TO_INSIDE" [0x54c7fa80, 0x0]

This change adds an optional match for the literal word protocol.

nhuff commented:
Should have a signed CLA now.

sathieu commented:
Anyone merging patches here?

sathieu commented on patterns/firewalls:
```diff
@@ -22,7 +22,7 @@ CISCOFW106015 %{CISCO_ACTION:action} %{WORD:protocol} \(%{DATA:policy_id}\) from
 CISCOFW106021 %{CISCO_ACTION:action} %{WORD:protocol} reverse path check from %{IP:src_ip} to %{IP:dst_ip} on interface %{GREEDYDATA:interface}
 # ASA-4-106023
-CISCOFW106023 %{CISCO_ACTION:action} %{WORD:protocol} src %{DATA:src_interface}:%{IP:src_ip}(/%{INT:src_port})?(\(%{DATA:src_fwuser}\))? dst %{DATA:dst_interface}:%{IP:dst_ip}(/%{INT:dst_port})?(\(%{DATA:dst_fwuser}\))?( \(type %{INT:icmp_type}, code %{INT:icmp_code}\))? by access-group %{DATA:policy_id} \[%{DATA:hashcode1}, %{DATA:hashcode2}\]
+CISCOFW106023 %{CISCO_ACTION:action}( protocol)? %{WORD:protocol} src %{DATA:src_interface}:%{DATA:src_ip}(/%{INT:src_port})?(\(%{DATA:src_fwuser}\))? dst %{DATA:dst_interface}:%{DATA:dst_ip}(/%{INT:dst_port})?(\(%{DATA:dst_fwuser}\))?( \(type %{INT:icmp_type}, code %{INT:icmp_code}\))? by access-group "?%{DATA:policy_id}"? \[%{DATA:hashcode1}, %{DATA:hashcode2}\]
```
This is %{IP:src_ip} instead of %{DATA:src_ip} and %{IP:dst_ip} instead of %{DATA:dst_ip}.

purbon commented:
Thanks a lot for your contribution @nhuff, it would be super nice if you can add test for this change. We try to enforce that all PR introduce some kind of testing, so we're able to increase the overall quality. Don't hesitate to ask any question regarding you might have, more than looking forward to help.

On the other side, because it took so long to visit this PR (I'm sorry for it), you will need to rebase it. 

What do you think about @sathieu comments? 

/cheers

nhuff commented:
Rebased to current master and added tests.  The reason I changed IP to DATA for the src and destination IPs is that I have examples of the FWSM logging hostnames in those fields instead of IPs.  I have a test that covers the case I have seen.

purbon commented:
LGTM
